angr/exploration_techniques/arm_system_explorer.py

```diff
@@ -0,0 +1,613 @@
+        This emulates the SVC instruction.
+
+        :param state: the current state.
+        :param insn : the SVC instruction.
+
+        :type state : angr.sim_state.SimState
+        :type insn  : angr.block.CapstoneInsn
+        """
+
+        state._ip += insn.size
+
+        if not ARMSystemExplorer._condition_passed(state, insn):
+            return
+
+        if state.se.eval(state.regs.itstate):
+            state.regs.itstate = state.se.LShR(state.regs.itstate, 8)
+
+        # The return address cannot be passed directly to the exception
+        # handler, since it may be off by 1 in Thumb mode (I blame this on
+        # libVEX...). Furthermore, this return address is used to compute the
+        # syscall number, so it has to be an word-aligned address.
+        #
+        # The odd address will be recomputed at exception return thanks to
+        # EPSR.T bit.
+        self._exception_entry(state, exception='SVCall', ret_addr=state._ip & ~1)
+
+    def _hook_cps(self, state, insn):
+        """
+        This emulates the CPS instruction.
+
+        :param state: the current state.
+        :param insn : the CPS instruction.
+
+        :type state : angr.sim_state.SimState
+        :type insn  : angr.block.CapstoneInsn
+        """
+
+        state._ip += insn.size
+
+        if not ARMSystemExplorer._condition_passed(state, insn):
+            return
+
+        if self._current_mode_is_privileged(state):
+            if insn.mnemonic.endswith('e'): # Interrupt Enable
+                if 'i' in insn.op_str:
+                    state.regs.primask &= ~1
+                if 'f' in insn.op_str:
+                    state.regs.faultmask &= ~1
+            elif insn.mnemonic.endswith('d'): # Interrupt Disable
+                if 'i' in insn.op_str:
+                    state.regs.primask |= 1
+                if 'f' in insn.op_str:
+                    state.regs.faultmask |= 1
+
+
+    def _hook_msr(self, state, insn):
+        """
+        Move to Special register from Register moves the value of a
+        general-purpose register to the selected special-purpose register.
+
+        :param state: the current state.
+        :param insn : the MSR instruction.
+
+        :type state : angr.sim_state.SimState
+        :type insn  : angr.block.CapstoneInsn
+        """
+
+        state._ip += insn.size
+
+        if not ARMSystemExplorer._condition_passed(state, insn):
+            return
+
+        sysm, rn = [str(i).lower() for i in insn.op_str.split(', ')]
+        privileged = self._current_mode_is_privileged(state)
+
+        l.debug("%s = %s" % (rn, repr(state.regs.__getattr__(rn))))
+
+        rn = state.regs.__getattr__(rn)
+
+        # xPSR
+        if sysm.endswith('psr'):
+            if 'a' in sysm:
+               state.regs.flags = rn
+            # Writes to IPSR and EPSR are ignored.
+
+        elif privileged:
+            # MSP/PSP
+            if sysm.endswith('sp'):
+                state.regs.__setattr__(sysm, rn)
```

Comment:
I would use `state.registers.store(sysm, rn)` rather than the setattr way. You're basically reducing it down to the former case anyways.